### PR TITLE
refactor: Change private variable to private constant

### PR DESCRIPTION
**Pull request by aragon999**
### 1. Why is this change necessary?
Only some minor changes, I stumbled upon, when going through the code, in particular moving the associations from a readonly private variable to a constant.

I hope in that case that no changelog is required. If otherwise, I will create one.

### 2. What does this change do, exactly?
Remove some lines of code.

### 3. Describe each step to reproduce the issue or behaviour.
\-

### 4. Please link to the relevant issues (if any).
\-

### 5. Checklist

- [x] I have rebased my changes to remove merge conflicts
- [ ] I have written tests and verified that they fail without my change
- [ ] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

### CODE REVIEW

File: src/Core/Checkout/Promotion/Gateway/PromotionGateway.php [hunk 1]
```diff
@@ -34,9 +36,6 @@ public function get(Criteria $criteria, SalesChannelContext $context): EntityCol
             new FieldSorting('priority', FieldSorting::DESCENDING)
         );
 
-        /** @var PromotionCollection $entities */
-        $entities = $this->promotionRepository->search($criteria, $context->getContext())->getEntities();
-
-        return $entities;
+        return $this->promotionRepository->search($criteria, $context->getContext())->getEntities();
```
Comment (mitelg) [inline src/Core/Checkout/Promotion/Gateway/PromotionGateway.php:43]: couldn't we change the return annotation to `PromotionCollection` :thinking: 

and at the same deprecating it, so it returns `PromotionCollection` with 6.7

Comment (aragon999) [reply src/Core/Checkout/Promotion/Gateway/PromotionGateway.php:43]: Yes, I was also thinking of this. But I was not sure regarding backwards compatibility. But when thinking about it in more detail, it should be fine, if we change the return type of the implementation (`PromotionGateway`) to `PromotionCollection`, as it still satisfies the interface, and add a deprecation notice on the interface (`PromotionGatewayInterface`), that the return type will be changed to `PromotionCollection`? :thinking: 

Would that be fine? Or do we need an actual implementation where the interface is changed to return a `PromotionCollection`?

Comment (mitelg) [reply src/Core/Checkout/Promotion/Gateway/PromotionGateway.php:43]: your idea sounds good :+1:

Comment (aragon999) [reply src/Core/Checkout/Promotion/Gateway/PromotionGateway.php:43]: I updated the PR accordingly :-) What do you think?

I now also added a changelog.

Not sure why the CI fails though, it should not be related to the changes.

**Review by mitelg — APPROVED** [review]

**Comment by marcelbrode** [comment]
Looks not bad! But it would be great to get the tests green 😄! Can you do that?

**Comment by mitelg** [comment]
> Looks not bad! But it would be great to get the tests green 😄! Can you do that?

a rebase should do the trick :slightly_smiling_face:

**Comment by aragon999** [comment]
I will do the rebase later, I am currently not at the PC.

Done that, but I am not sure about this error: https://github.com/shopware/shopware/actions/runs/11222110933/job/31193869990?pr=4931#step:7:800

Is that fine, or should that be changed? Is it, because this class is not final? :thinking:

**Comment by marcelbrode** [comment]
Ah, that check says it's breaking. I guess you have to deprecate it there as well, as you did in some of your other files

**Comment by aragon999** [comment]
Yes, the question is, why it says that it is breaking. As this does not break when one properly implements the interface. But I now reverted that change. But I added the final annotation (which should not break, as it is only an annotation).

The now failing OpenApi check is also failing on trunk.

**Comment by marcelbrode** [comment]
> Yes, the question is, why it says that it is breaking. As this does not break when one properly implements the interface. But I now reverted that change. But I added the final annotation (which should not break, as it is only an annotation).
> 
> The now failing OpenApi check is also failing on trunk.

Well, that's a static check, so _in theory_ the return type is now more limited.
Maybe @keulinho can give more insights about what is fine and what not.

File: src/Core/Checkout/Promotion/Gateway/PromotionGateway.php [hunk 2]
```diff
@@ -25,7 +29,9 @@ public function __construct(private readonly EntityRepository $promotionReposito
      * Gets a list of promotions for the provided criteria and
      * sales channel context.
      *
-     * @return EntityCollection<PromotionEntity>
+     * @deprecated tag:v6.7.0 - Return type will be natively changed to PromotionCollection
```
Comment (mitelg) [inline src/Core/Checkout/Promotion/Gateway/PromotionGateway.php]:
```suggestion
     * @deprecated tag:v6.7.0 - reason:return-type-change - Return type will be changed to PromotionCollection
```

with this PHPStan will not complain

File: changelog/_unreleased/2024-10-02-changed-promotiongatewayinterface-return-type-to-promotioncollection.md [hunk 3]
```diff
@@ -0,0 +1,17 @@
+---
+title: Changed `PromotionGatewayInterface` return type to `PromotionCollection`
+issue: NEXT-0000
+author: Max
+author_email: max@swk-web.com
+author_github: @aragon999
+---
+# Core
+* Changed the return type of the `Shopware\Core\Checkout\Promotion\Gateway\PromotionGatewayInterface` from `Shopware\Core\Framework\DataAbstractionLayer\EntityCollection<PromotionEntity>` to `Shopware\Core\Checkout\Promotion\PromotionCollection`, which will be adjusted in the next major Shopware version
+* Changed the return type of `Shopware\Core\Checkout\Promotion\Gateway\PromotionGateway` to `Shopware\Core\Checkout\Promotion\PromotionCollection`
```
Comment (mitelg) [inline changelog/_unreleased/2024-10-02-changed-promotiongatewayinterface-return-type-to-promotioncollection.md]: the changelog also needs to be adjusted according to the latest changes

File: src/Core/Checkout/Promotion/Gateway/PromotionGateway.php (hunk 2 above)
Comment (aragon999) [reply src/Core/Checkout/Promotion/Gateway/PromotionGateway.php]: Done, thank you :-)

**Comment by aragon999** [comment]
@marcelbrode yes in general the implementation of an interface (or subclass in that case), can narrow down the return type, i.e. see: https://3v4l.org/c6bse

The only problem one can run into, is if one directly extended the `PromotionGateway` and does not change the return type. So I think it should be the least breaking behavior, which is now implemented, in case anybody directly extended the `PromotionGateway` which should in future not be possible anymore, hence the `@final` doctype.

**Comment by marcelbrode** [comment]
I guess that's fine. I'll import it 🙂

**Comment by marcelbrode** [comment]
Your Pull Request has been merged! :tada:
Thanks for your contribution! :blue_heart:

**Pull request closed without merging**